Structured CBOR. Fixes #2975, fixes #2848, fixes #3143

JesusMcCloud opened the pull request:
fixes #2975, #2848

This PR introduces structured CBOR encoding and decoding

## Encoding from/to `CborElement`

Bytes can be decoded into an instance of `CborElement` with the [Cbor.decodeFromByteArray] function by either manually
specifying [CborElement.serializer()] or specifying [CborElement] as generic type parameter.  
It is also possible to encode arbitrary serializable structures to a `CborElement` through [Cbor.encodeToCborElement].

Since these operations use the same code paths as regular serialization (but with specialized serializers), the config flags
behave as expected

## Newly introduced CBOR-specific structures

* [CborPrimitive] represents primitive CBOR elements, such as string, integer, float boolean, and null.
  CBOR byte strings are also treated as primitives  
  Each primitive has a [value][CborPrimitive.value]. Depending on the concrete type of the primitive, it maps
  to corresponding Kotlin Types such as `String`, `Int`, `Double`, etc.
  Note that Cbor discriminates between positive ("unsigned") and negative ("signed") integers!  
  `CborPrimitive` is itself an umbrella type (a sealed class) for the following concrete primitives:
  * [CborNull] mapping to a Kotlin `null`
  * [CborBoolean] mapping to a Kotlin `Boolean`
  * [CborInt] which is an umbrella type (a sealed class) itself for the following concrete types
    (it is still possible to instantiate it as the `invoke` operator on its companion is overridden accordingly):
    * [CborPositiveInt] represents all `Long` numbers `≥0`
    * [CborNegativeInt] represents all `Long` numbers `<0`
  * [CborString] maps to a Kotlin `String`
  * [CborFloat] maps to Kotlin `Double`
  * [CborByteString] maps to a Kotlin `ByteArray` and is used to encode them as CBOR byte string (in contrast to a list
    of individual bytes)

* [CborList] represents a CBOR array. It is a Kotlin [List] of `CborElement` items.

* [CborMap] represents a CBOR map/object. It is a Kotlin [Map] from `CborElement` keys to `CborElement` values.
  This is typically the result of serializing an arbitrary 

# Example

```hexdump
bf                                 # map(*)
   61                              #   text(1)
      61                           #     "a"
   cc                              #   tag(12)
      1a 0fffffff                  #     unsigned(268,435,455)
   d8 22                           #   base64 encoded text, tag(34)
      61                           #     text(1)
         62                        #       "b"
                                   #     invalid length at 0 for base64
   20                              #   negative(-1)
   d8 38                           #   tag(56)
      61                           #     text(1)
         63                        #       "c"
   d8 4e                           #   typed array of i32, little endian, twos-complement, tag(78)
      42                           #     bytes(2)
         cafe                      #       "\xca\xfe"
                                   #     invalid data length for typed array
   61                              #   text(1)
      64                           #     "d"
   d8 5a                           #   tag(90)
      cc                           #     tag(12)
         6b                        #       text(11)
            48656c6c6f20576f726c64 #         "Hello World"
   ff                              #   break
```

Decoding it results in the following CborElement (shown in manually formatted diagnostic notation):

```
CborMap(tags=[], content={  
    CborString(tags=[],   value=a) = CborPositiveInt( tags=[12],     value=268435455),  
    CborString(tags=[34], value=b) = CborNegativeInt( tags=[],       value=-1),  
    CborString(tags=[56], value=c) = CborByteString(  tags=[78],     value=h'cafe),  
    CborString(tags=[],   value=d) = CborString(      tags=[90, 12], value=Hello World)  
})
```

## Implementation Details
I tried to stick to the existing CBOR codepaths as closely as possible, and the approach to add tags directly to CborElements is the most pragmatic way of getting expressiveness and convenient use. It does come with a caveat (also taken from the Readme:

Tags are properties of `CborElements`, and it is possible to mixing arbitrary serializable values with `CborElement`s that
contain tags inside a serializable structure. It is also possible to annotate any [CborElement] property
of a generic serializable class with `@ValueTags`.  
**This can lead to asymmetric behavior when serializing and deserializing such structures!**

The test cases (and comments in the test cases reflect this

## Closing Remarks
I also fixed a faulty hex input test vector that I introduced myself, last year, if I pieced it together correctly (see [here](https://github.com/Kotlin/kotlinx.serialization/pull/3036#issuecomment-3163933606)) and I amended the benchmarks. (see [here](https://github.com/Kotlin/kotlinx.serialization/pull/3036#issuecomment-3163686974)).

Since the commits from here will be squashed anyways, I did not care for a clean history.

JesusMcCloud committed:
WIP strucutred cbor

JesusMcCloud committed:
proper when for decoding

JesusMcCloud committed:
baseline tagging

JesusMcCloud committed:
some cleanups

pdvrieze commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:
```diff
@@ -34,12 +34,12 @@ public sealed class Cbor(
-                verifyKeyTags = false,
-                verifyValueTags = false,
-                verifyObjectTags = false,
+                encodeKeyTags = true,
+                encodeValueTags = true,
+                encodeObjectTags = true,
+                verifyKeyTags = true,
+                verifyValueTags = true,
+                verifyObjectTags = true,
```
As these parameters change the behaviour of the system you should not change the default. Instead you could provide a way for users to choose a newer/different configuration.

JesusMcCloud replied:
thanks! definitely not intended

pdvrieze commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,202 @@
+     * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+     */
+    @OptIn(ExperimentalUnsignedTypes::class)
+    public val tags: ULongArray = ulongArrayOf()
```
You probably want to keep out the default initializer to ensure that any concrete subclass passes the value (it is sealed so it is not a usability issue to require it to be passed along).

pdvrieze replied:
Note that if tags are uncommon you should probably make it nullable to avoid the overhead of the empty array. Alternatively you could create an internal empty array value that is reused.

JesusMcCloud replied:
i'd go for an internal empty array. It does not make sense to differentiate between a null value and an empty array

pdvrieze commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,202 @@
+@Serializable(with = CborIntSerializer::class)
+public class CborNegativeInt(
+    public val value: Long,
+    tags: ULongArray = ulongArrayOf()
```
See above, remove the default

pdvrieze commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,202 @@
+ * Class representing CBOR `null` value
+ */
+@Serializable(with = CborNullSerializer::class)
+public class CborNull(tags: ULongArray=ulongArrayOf()) : CborPrimitive(tags) {
```
The default value should remain, but use an internal array (or null) to avoid extensive copies of empty arrays.

pdvrieze commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/CborElementSerializers.kt:
```diff
@@ -0,0 +1,304 @@
+        }
+
+    override fun serialize(encoder: Encoder, value: CborElement) {
+        encoder.asCborEncoder()
```
Ideally you allow serializing to other formats with the "normal" behaviour (see the serializers for `JsonElement` etc.)

JesusMcCloud replied:
doesn't the JsonElement serializer have a `verify` method that ensures it can only be used with the JSON format?

JesusMcCloud replied:
turns out: [it does](https://github.com/Kotlin/kotlinx.serialization/blob/10c7b9d2d6a333aee66c15c2f8d224b8e7c71660/formats/json/commonMain/src/kotlinx/serialization/json/JsonElementSerializers.kt#L193)

pdvrieze replied:
@JesusMcCloud You are right. I had missed the check because it was encapsulated. The worst part is that it would actually work without that verify check.

JesusMcCloud replied:
it does make life easier for those implementing the serializers. for me at least, i don't have to think about how this will behave with other formats, esp. wrt. tagging.

pdvrieze commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/CborElementSerializers.kt:
```diff
@@ -0,0 +1,304 @@
+    }
+}
+
+public object CborIntSerializer : KSerializer<CborNegativeInt> {
```
Should be internal like the other specialist serializers. They are accessible on the type itself.

pdvrieze commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,202 @@
+ * Class representing signed CBOR integer (major type 1).
+ */
+@Serializable(with = CborIntSerializer::class)
+public class CborNegativeInt(
```
The name NegativeInt is misleading as it appears to mean a value that can only be negative, not a signed value (that can be positive or negative). Given this name is poor, you also want to rename the counterpart from PositiveInt.

JesusMcCloud replied:
That was intentional. IIRC only negative ints are encoded as signed int in cbor, so it really must be a negative int

pdvrieze replied:
But what if they are created programmatically? If you have this value based type you need a factory that creates the correct version on demand (and probably have an "CborInt" parent type that is sign independent)

JesusMcCloud replied:
a sealed parent makes sense (with a custom invoke function that returns the subtype based on sign), but the name for this class is the only one i could think of that is actually accurate

JesusMcCloud replied:
it definitely makes life easier when you don't parse, but create. when parsing, though, I am slightly leaning towards being very explicit and very close to the wire format (which is what we're representing). I'll let @sandwwraith decide

pdvrieze commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/CborElementSerializers.kt:
```diff
@@ -0,0 +1,304 @@
+ * Used to resolve cyclic dependencies between recursive serializable structures.
+ */
+@OptIn(ExperimentalSerializationApi::class)
+private fun defer(deferred: () -> SerialDescriptor): SerialDescriptor = object : SerialDescriptor {
+    private val original: SerialDescriptor by lazy(deferred)
+
+    override val serialName: String
+        get() = original.serialName
+    override val kind: SerialKind
+        get() = original.kind
+    override val elementsCount: Int
+        get() = original.elementsCount
+
+    override fun getElementName(index: Int): String = original.getElementName(index)
+    override fun getElementIndex(name: String): Int = original.getElementIndex(name)
+    override fun getElementAnnotations(index: Int): List<Annotation> = original.getElementAnnotations(index)
+    override fun getElementDescriptor(index: Int): SerialDescriptor = original.getElementDescriptor(index)
+    override fun isElementOptional(index: Int): Boolean = original.isElementOptional(index)
+}
```
Rather than this, it may be better to have the descriptor property itself be lazy, rather than each of its elements.

pdvrieze reviewed (commented):
I've reviewed the code at a general level. There is a lot of repetition, so I've only commented on the first case, not every one. I guess some of the AI generation is visible in the details.

JesusMcCloud committed:
cleanups

JesusMcCloud committed:
tree encoding half-done

JesusMcCloud committed:
tree encoding close

JesusMcCloud committed:
polish encoder

JesusMcCloud committed:
visibility fixes

JesusMcCloud committed:
more checks

JesusMcCloud committed:
WIP decode from CborElement

JesusMcCloud committed:
more AI slop

JesusMcCloud committed:
cleanup after Junie

JesusMcCloud committed:
clean up more

JesusMcCloud committed:
fix structural issues

JesusMcCloud committed:
fix map size regression

JesusMcCloud committed:
streamlining and cleanups

JesusMcCloud committed:
benchmarks

JesusMcCloud committed:
add more tests

JesusMcCloud committed:
clarify faulty test vector

JesusMcCloud committed:
fix test vector

JesusMcCloud committed:
Fix Tagging and Simplify

JesusMcCloud committed:
finalize api

JesusMcCloud committed:
APIDUMP

JesusMcCloud committed:
docs + apidump

whyoleg reviewed (commented):
First round of review from my side :)
Mostly reviewed API surface, will take a look on the implementation details later

JesusMcCloud commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,244 @@
+ * traditional methods like [List.get] or [List.size] to obtain CBOR elements.
+ */
+@Serializable(with = CborListSerializer::class)
+public class CborList(
```
Sadly, there is an annotation by the very same name ins the same package. Now moving the cbor elements is easily done, but it will probably not be very intuitive (thinking of autocompletion here).

In any case: Good point but not my call!

whyoleg replied:
Oh, yes, how did I miss that :(

JesusMcCloud replied:
`@EncodeAsCborArray`?

JesusMcCloud commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,244 @@
+     * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+     */
+    @OptIn(ExperimentalUnsignedTypes::class)
+    public var tags: ULongArray = tags
```
We can make it so that there is a val with a custom getter that delegates to the actual field, but the tags should still be simply written to a CborElement as they are collected upon deserialisation, to avoid serialization overhead. Actually, the `elements` should probably be mutable too internally so the whole container structure that now collects tags and elements for Cbor structures can go away.

In the end this would mean that we have:

- an `internal var` making it possible to collect tags during deserialisation, but a `public val tags` with a custom getter and no backing field
- an `internal val elements` for array and map so we can also collect elements as we go and get rid of another instantiation, but a `public val elements: List<CborElement>`

This should boost performance for deserialisation significantly

JesusMcCloud replied:
- `val` is fine for the public API
- tags are `ULong`s according to spec
- arrays in Kotlin are a cursed as they are in Java

How about just exposing a `List<ULong>` as part of the public API? we could then still mutate as part of the deserialization process internally and not expose that mutability

sandwwraith replied:
They will be boxed, though. Theoretically, we can make a defensive copy in a custom getter instead, but I'm not sure it will be faster (less memory pressure tho).
Also, we have already gone down the road with experimentality with `@KeyTags/ValueTags`.

fzhinkin replied:
> Also, we have already gone down the road with experimentality with `@KeyTags/ValueTags`.

True. Yet, the mutability concern holds still (and there's a similar one for `CborByteString`, BTW), so we need some other way to represent tags than by using an array.

fzhinkin replied:
> less memory pressure tho

Speaking of memory pressure, the current implementation allocates a fresh `ULongArray` instance for every `CborElement`. Assuming that majority of element won't have tags associated with them, it feels pretty excessive.

JesusMcCloud commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/CborElementSerializers.kt:
```diff
@@ -0,0 +1,329 @@
+    )
+
+/*need to expose writer to access encodeTag()*/
+internal fun Encoder.asCborEncoder() = this as? CborWriter
```
You are probably on to something here. I should investigate tag handling some more

JesusMcCloud replied:
done in 1d97928929c33cc9db117c8df3ff56a599bab8dc

JesusMcCloud commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,261 @@
+@Serializable(with = CborPositiveIntSerializer::class)
+public class CborPositiveInt(
+    value: ULong,
+    vararg tags: ULong
```
I agree with the first part, but the second part is simply in line with how the tagging annotations behave and it is convenient for the user (much more so that `ulongArrrayOf()`.
Yet, I am aware that it should be consistent, so making ints work differently from everything else is also not really nice… In the end, don't really have an opinion. Sort it out internally, it's a straight-forward refactor anyways.

What would help in such cases is a language feature that forces parameter name specification at the call site, but that is not going to happen anytime soon, if ever.

JesusMcCloud replied:
Those two features would help, but in the meantime, we need something else for a solution

JesusMcCloud commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,244 @@
+    }
+}
+@Serializable(with = CborIntSerializer::class)
+public sealed class CborInt<T : Any>(
```
Good catch! I agree. The constructors with the current signature (simply passing a number) should stay for convenience.

JesusMcCloud replied:
done in 89d671c2cd280e503f9ea6728fbe2683c8c88001

JesusMcCloud commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,261 @@
+ * The whole hierarchy is [serializable][Serializable] only by [Cbor] format.
+ */
+@Serializable(with = CborElementSerializer::class)
+public sealed class CborElement(
```
The first one solves the issues nicely **and** it would work well with the internally mutable tags array s.t. it would avoid instantiation overhead just for modifying tags. BUT, that would be a footgun for the consumer.

I have thought about CborTagged, also because of the asymmetric serialization/deserialization behaviour, which really is not all that nice, but also very niche (why would anyone in their right mind do this??)

When I started with the implementation, I was imagining how the most straight-forward and usable way would look like: Tags are attached to something and have a hierarchy. An array is enough to represent this and makes for a nice API that mirrors the pattern of `@ValueTags`, `@ObjectTags`, and `@KeyTags`. Now you could get rid of that pesky asymmetry, if you disallow those annotations on `CborElement`. Then I'd assume we'd be back to straightforward logic again?!

Bottom line: there are issues that need sorting out. The inconsistency issue is documented in the readme, so I was not trying to cover anything up, but I wanted unbiased opinions and fresh ideas because I could not get to a satisfactory solution by myself. It worked once already. I just hope Leonid checks the code first, and the comments later ;-)

whyoleg commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborDecoder.kt:
```diff
@@ -31,4 +31,15 @@ public interface CborDecoder : Decoder {
+     * This method is allowed to invoke only as the part of the whole deserialization process of the class,
+     * calling this method after invoking [beginStructure] or any `decode*` method will lead to unspecified behaviour.
+     */
+    public fun decodeCborElement(): CborElement
```
Most probably to `encodeSerializableValue(CborElementSerializer, element)`, but yes, just for symmetry

JesusMcCloud replied:
done in 004dd1dd569a517604bc3272a8793d544f0a105e

JesusMcCloud committed:
refactor cborint

JesusMcCloud committed:
add CborEncoder.encoderCborElement

JesusMcCloud committed:
pimp CborEncoder

JesusMcCloud commented:
Any updates on the open discussion points?

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,300 @@
+ */
+@Serializable(with = CborIntSerializer::class)
+@ExperimentalSerializationApi
+public class CborInt(
```
What do you think about `CborInteger` instead? It's closer to major type descriptions from the spec (`an unsigned integer`, `a negative integer`) and it is farther from Kotlin's `Int`.

JesusMcCloud replied:
I'm fine with both. Completely up to you guys

fzhinkin replied:
Let's make it `CborInteger` then.

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,300 @@
+    }
+
+    /**
+     * **WARNING! Possible truncation/overflow!** E.g., `-2^64` -> `1`
```
Silent truncation/overflow is something we should avoid. I'd vote for an exception if a value could not be represented as `Long` (`Int`, `Short`, or `Byte` - more on that later), and an additional `toLongOrNull` function, that'll return `null` on overflow.

It would be nice to align the API with what we already have for `JsonElement` and instead of providing a member function, provide extension properties.
And it seems reasonable to provide properties returning other integer types too.

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,300 @@
+            ")"
+    }
+
+}
```
There are few things that are currently missing, but that would otherwise improve DX when it comes to using CborElement API:
- casting functions/properties such as
  - `val CborElement.cborFloat: CborFloat`, `val CborElement.cborList: CborList`
   - `fun CborElement.asCborFloat(): CborFloat`, ...
- builders for aggregates, similar to `JsonObjectBuilder` and `JsonArrayBuilder`

Those could be added later, though.

fzhinkin reviewed (changes requested):
@JesusMcCloud, sorry for the delay with review. 😿 

I didn't thoroughly review the serialization machinery, but focused on the API part for now.

JesusMcCloud commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborEncoder.kt:
```diff
@@ -31,4 +34,48 @@ public interface CborEncoder : Encoder {
+    public fun encodeCborElement(element: CborElement): Unit = encodeSerializableValue(CborElementSerializer, element)
+
+    /**
+     * Allows manually encoding CBOR tags. Use with caution, as it is possible to produce invalid CBOR if invoked carelessly!
```
From the top of my head (without reading up again in the CBOR spec), I'm afraid this might be far from trivial.

JesusMcCloud commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,300 @@
+@ExperimentalSerializationApi
+public class CborInt(
+    absoluteValue: ULong,
+    public val sign: Sign,
```
It is simpler, but it's also what is being done in virtually every integer implementation that requires a dedicated signum. I don't really care one way or the other, so I'll let you sort this one out with @whyoleg.

fzhinkin replied:
IIRC @whyoleg was fine with a boolean flag; Oleg, right? 😈

whyoleg replied:
Yes! I'm fine with a flag like `isPositive`/`isNegative`.
Should we also then add an assertion that if the value is `0`, we should have `isPositive=true`?

JesusMcCloud commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,300 @@
+@Serializable(with = CborNullSerializer::class)
+@ExperimentalSerializationApi
+public class CborNull(vararg tags: ULong) : CborPrimitive<Unit>(Unit, tags)
+
```
Great catch! We need a `CborUndefined`. I totally forgot about that!

JesusMcCloud committed:
Apply suggestions from code review

Co-authored-by: Filipp Zhinkin <filipp.zhinkin@gmail.com>

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/CborElementSerializers.kt:
```diff
@@ -0,0 +1,300 @@
+
+    override fun deserialize(decoder: Decoder): CborFloat {
+        decoder.asCborDecoder()
+        return CborFloat(decoder.decodeDouble())
```
Tags will be lost here during deserialization (the same is true for CborMap and CborList deserialization routines).

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/CborElementSerializers.kt:
```diff
@@ -0,0 +1,300 @@
+        val cborEncoder = encoder.asCborEncoder()
+        cborEncoder.encodeTags(value.tags)
+        //this we really don't want to expose so we cast here
+        (cborEncoder as CborWriter).encodeByteString(value.value)
```
But the `CborEncoder` is public and it could be anything else.

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/CborElementSerializers.kt:
```diff
@@ -0,0 +1,300 @@
+        }
+
+        decoder.decodeNull()
+        return CborNull()
```
Tags will be lost here too

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoder.kt:
```diff
@@ -177,10 +182,207 @@ internal class IndefiniteLengthCborWriter(cbor: Cbor, output: ByteArrayOutput) :
+            is Map -> CborMap(
+                content = if (elements.isNotEmpty()) IntRange(0, elements.size / 2 - 1).associate {
+                    elements[it * 2] to elements[it * 2 + 1]
+                } else mapOf(),
```
I would also vote for 
```kotlin
buildMap {
    for (i in 0 until elements.size / 2 - 1) {
        put(elements[i * 2], elements[i * 2 + 1])
     }
}
```
as an alternative to `if (..) associate {} or mapOf()`

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoder.kt:
```diff
@@ -177,10 +182,207 @@ internal class IndefiniteLengthCborWriter(cbor: Cbor, output: ByteArrayOutput) :
+        class List(tags: ULongArray, elements: MutableList<CborElement> = mutableListOf()) :
+            CborContainer(tags, elements)
+
+        class Primitive(tags: ULongArray) : CborContainer(tags, elements = mutableListOf()) {
```
`Primitive` is unused and could be removed, isn't it?

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoder.kt:
```diff
@@ -177,10 +182,207 @@ internal class IndefiniteLengthCborWriter(cbor: Cbor, output: ByteArrayOutput) :
+
+    // value tags are collects inside beginStructure, so we need to cache them here and write them in beginStructure or encodeXXX
+    // and then null them out, so there are no leftovers
+    private var nextValueTags: ULongArray = ulongArrayOf()
```
It's better to be a function with a name that'll give away the behavior.

Also, I'm not sure that reallocating array every time we need to append a tag is the best way to collect tags. `ProtobufTaggedBase` shows a more efficient approach, although for CBOR `tagsStack`-analogue should be nullable and has `null` value by default (I assume that elements are usually untagged).

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/Cbor.kt:
```diff
@@ -88,7 +88,49 @@ public sealed class Cbor(
+ * Deserializes the given [element] element into a value of type [T] using a deserializer retrieved
+ * from reified type parameter.
+ *
+ * @throws [SerializationException] if the given JSON element is not a valid CBOR input for the type [T]
```
I think there are no such JsonElements that are valid CBOR inputs ;)

whyoleg commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborEncoder.kt:
```diff
@@ -31,4 +34,48 @@ public interface CborEncoder : Encoder {
+    /**
+     * Encode a positive value as [CborInt]. This function exists to encode negative values exceeding [Long.MAX_VALUE]
+     */
+    public fun encodePositive(value: ULong)
```
It feels a bit strange that we expose those.
If users need those, they should be able to just do:
```kotlin
encodeCborElement(CborInteger(...))
```

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Decoder.kt:
```diff
@@ -531,14 +570,210 @@ private fun Iterable<ByteArray>.flatten(): ByteArray {
+    override fun isNull() =
+        if (layer.isStructure) layer.peek().let {
+            it is CborNull ||
+                /*THIS IS NOT CBOR-COMPLIANT but KxS-proprietary handling of nullable classes*/
```
Could you please clarify what do you mean by "non-compliant default behavior"? Null objects being represented by empty maps?

JesusMcCloud commented:
Note to self: also provide examples and test map delegation with non-string keys

JesusMcCloud committed:
CBOR: Optionally encode nullable complex props as empty map (fixes #2848)

JesusMcCloud committed:
cborInt sign refactor

JesusMcCloud committed:
Fix structured CBOR element encoding

JesusMcCloud committed:
Expose byte string and undefined encoding

JesusMcCloud committed:
Adjust null object encoding expectations

JesusMcCloud committed:
Reuse shared empty tags array

JesusMcCloud committed:
Update CBOR API dumps

JesusMcCloud committed:
Use shared empty tags in decoder

JesusMcCloud committed:
Fix encodePositive KDoc

JesusMcCloud committed:
Remove encodePositive/encodeNegative from CborEncoder

JesusMcCloud committed:
Update CBOR API dumps

JesusMcCloud committed:
Document and harden large integer encoding

JesusMcCloud committed:
Restore encodePositive/encodeNegative API

JesusMcCloud committed:
Make CborPrimitive non-generic

JesusMcCloud committed:
Add test for encodeCborElement misuse

JesusMcCloud committed:
Revert "Add test for encodeCborElement misuse"

This reverts commit 847be408efb039714a5130a93f8c1cbb890f24d2.

JesusMcCloud committed:
Add top-level CborInt factory functions

JesusMcCloud committed:
Add safe CborInt numeric conversions

JesusMcCloud committed:
Fix StructuredCborWriter pending tag handling

JesusMcCloud committed:
Fix CBOR container

JesusMcCloud committed:
Test structured CBOR root primitives and lists

JesusMcCloud committed:
Fix typed CborElement decoding and add tests

JesusMcCloud committed:
Validate encodeTags is followed by a data item

JesusMcCloud committed:
CborInt -> CborInteger

JesusMcCloud committed:
Rename @CborArray to @CborObjectAsArray

JesusMcCloud committed:
Rename CborList -> CborArray

JesusMcCloud committed:
Fix CBOR after dev rebase

- Align byte parser/skip logic with dev (length checks, truncation, indefinite chunks, error messages)\n- Keep structured CBOR helpers warning-free under -Werror\n- Restore tree reader EOF semantics for CborElement decoding

JesusMcCloud committed:
Remove copyright headers from new CBOR files

JesusMcCloud committed:
Add test documenting unsigned CBOR interop issue

JesusMcCloud committed:
CBOR: encode/decode Kotlin unsigned types as CBOR unsigned ints

Handle kotlin.UByte/UShort/UInt/ULong inline serializers via format-specific
encodeInline/decodeInline wrappers so values use CBOR major type 0 and encode/decode correctly.

JesusMcCloud committed:
CBOR: add map-backed delegated properties test

JesusMcCloud committed:
fix CborInt function name

fzhinkin commented:
> So that I get this right: This PR will get moved to another branch and we'll continue over there with me making smaller PRs to address issues you raise. So it will be multiple small, manageable PRs against a separate branch and then there will be a final Godzilla-sized PR against `dev` that won't require in-depth review any more.

That's correct. 

> The only point I wanted to raise was: If possible, please check the API surface first and comment on that. If multiple partial API changes cause too many moving parts on the underlying implementation that disagree, I'll be refactoring the same code sections multiple times for nought. 

Sure, the part of my proposal it to present a list of API issues that needs to be solved, so that we all will be on the same page and understand where things are going. 

I'll come back with such a list in the beginning of next week and then, once we agree on it, we can proceed further.

> Just ping me when there are concrete things to do. You can also reach me on slack if you want to discuss ideas without spamming GitHub.

Absolutely. Thanks!

JesusMcCloud committed:
CBOR: remove nullAsMap

JesusMcCloud committed:
remove abstract value from base CBOR element

JesusMcCloud committed:
CBOR: rearrange internal CBOR writer apis

JesusMcCloud committed:
CBOR: guard raw elements from tag/label annotations

also: respect CBOR config switches wrt. tags+labels

JesusMcCloud committed:
CBOR: update docs

JesusMcCloud committed:
CBOR: safer wrapping of mutable tags and bytestring + convenience copy methods+ctors

JesusMcCloud committed:
CBOR: Rename writer/parser class hierarchies

fzhinkin commented on formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborArrayTest.kt:
```diff
@@ -10,14 +10,18 @@ class CborArrayTest {
         /**
          * 81    # array(1)
-         *    26 # negative(6)
+         *    26 # negative(-7)
```
Minor nit: that's an output from cbor.me and it's worth keeping it as it is.

fzhinkin commented on formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborElementTest.kt:
```diff
@@ -0,0 +1,1237 @@
+                assertEquals(hex, cbor.encodeToHexString(struct))
+                assertEquals(struct, cbor.decodeFromHexString<CborElement>(hex))
+                //we have an ambiguity here (null vs. CborNull), so we cannot compare for equality with the object
+                //assertEquals(obj, cbor.decodeFromCbor(struct))
```
It would be nice to assert the current behavior anyway (and also add a test for tagged null value which will be decoded to `CborNull(tags = ...)`).

Also, JsonElement API behaves the same way (well, I mean nullable json element will be decoded as `null`).

fzhinkin commented on formats/cbor/commonTest/src/kotlinx/serialization/cbor/CborElementTest.kt:
```diff
@@ -0,0 +1,1237 @@
+
+    @Test
+    fun testDecodeFloatingPoint() {
+        // Test data from CborParserTest.testParseDoubles
```
Here's a peculiar issue around FP numbers encoding:
- when encoding a Float property directly to byte array, it will be encoded as float (0xFA followed by 4 bytes)
- however, if the object will be encoded to `CborElement` and then encoded to byte array, this property will be encoded as double (0xFB followed by 8 bytes).

1) I wonder if it's an issue
2) Perhaps, `CborFloat` should capture exact underlying representation if possible?

The problem would not exist for deterministic encoding (#3021).

JesusMcCloud replied:
Nice catch! I agree that `CborFloat` should remember/know how many bits of precision to use.

- When parsing, we want to accurately represent what was on the wire.
- When constructing, I am very much for canonicalising where we can, which means that should encode using 64 bit. The safe way to nudge the user towards this without taking away control, is probably to have the ctor that takes a double be regularly accessible and introduce a secondary ctor taking a float, but guard that using an opt-in annotation.

WDYT?

fzhinkin replied:
> When parsing, we want to accurately represent what was on the wire.

There's no way to represent half-floats at the moment, as we still don't have such type.
At the same time, a float can precisely represent all half-fp values, and a double can precisely represent all float values. So it should be enough to continue storing the value expanded to `Double` along with some enum value telling us about the original encoding.

> When constructing, I am very much for canonicalising where we can, which means that should encode using 64 bit. 

From the RFC standpoint, "canonical" meant "shortest representation possible" :)
- https://www.rfc-editor.org/info/rfc7049/#section-3.9
- https://www.rfc-editor.org/rfc/rfc8949.html#section-4.2.1 (it's no longer called "canonical", though)

> The safe way to nudge the user towards this without taking away control, is probably to have the ctor that takes a double be regularly accessible and introduce a secondary ctor taking a float, but guard that using an opt-in annotation.

If `CborFloat` will store a double value along with the "precision" attribute, then there indeed could be two constructors, one accepting `Double` and one accepting `Float` (and then the third one, accepting `Float16`, if we will add it to the stdlib or the language), and by knowing which one was invoked, we can assign the "precision" value and later use it for serialization.

In other words, something like:
```
enum class Precision { SINGLE, DOUBLE /*, HALF */ }

class CborFloat internal constructor(val value: Double, val precision: Precision) {
   public constructor (value: Double) : this(value, Precision.SINGLE)
   public constructor (value: Float): this(value.toDouble(), Precision.SINGLE)
}
```

fzhinkin commented on formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt:
```diff
@@ -0,0 +1,608 @@
+     * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+     */
+    @OptIn(ExperimentalUnsignedTypes::class)
+    tags: ULongArray = EMPTY_TAGS
```
CborElement's tags will not be encoded unless `encodeValueTags` is set to `true`. If that's an intended behavior, it has to be explicitly specified for both `tags` property (or `CborElement`) and for `encodeValueTags`.

It also seems a bit odd that decoded CborElements will have tags (if tags were in the decoded byte stream, of course) even if `encodeValueTags` is disabled.

fzhinkin reviewed (commented):
@JesusMcCloud, thank you for all the efforts you put into it! I think all main design questions are closed now (modulo the way tags are represented, but I'm a proponent of the current approach 😈).

There are few new comments from me, but I'm planning to merge this branch into a separate branch forked from the `dev` (as it was proposed some time ago) and continue with it. There are a few things I would love to fix before mering everything into the `dev`, so all other minor changes can be fixed via small PRs against that branch.
